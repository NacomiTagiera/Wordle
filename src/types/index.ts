export type GameStatus = 'fail' | 'playing' | 'win';

export type LetterState = 'absent' | 'correct' | 'present';

export type Message = {
  message: string;
  duration: number;
};

export type GameState = {
  boardRowState: LetterState[][];
  boardState: string[];
  currentRowIndex: number;
  keyboardLetterState: Record<string, LetterState>;
  message: Message;
  solution: string;
  status: GameStatus;
};
