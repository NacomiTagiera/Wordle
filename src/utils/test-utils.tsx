import React, { type PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { render, type RenderOptions } from '@testing-library/react';

import { type AppStore, type RootState, setupStore } from '@/redux/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
