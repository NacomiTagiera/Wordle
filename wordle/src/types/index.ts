export interface GameState {
  boardRowState: LetterState[][];
  boardState: string[];
  currentRowIndex: number;
  keyboardLetterState: Record<string, LetterState>;
  message: Message;
  solution: string;
  status: GameStatus;
}

export type GameStatus = "fail" | "playing" | "win";

export type LetterState = "absent" | "correct" | "present";

export interface Message {
  message: string;
  duration: number;
}
