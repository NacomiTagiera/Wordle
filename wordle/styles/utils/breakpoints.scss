$breakpoints-up: (
  "small": "415px",
  "medium": "768px",
  "large": "1024px",
);

@mixin breakpoint($size) {
  @media (min-width: map-get($breakpoints-up, $size)) {
    @content;
  }
}
