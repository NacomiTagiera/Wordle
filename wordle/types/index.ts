export interface GameState {
  boardState: string[];
  currentRowIndex: number;
  keyboardletterState: { [letter: string]: LetterState };
  solution: string;
  status: GameStatus;
}

export type GameStatus = "fail" | "playing" | "win";

export enum LetterState {
  Absent,
  Correct,
  Present,
}
