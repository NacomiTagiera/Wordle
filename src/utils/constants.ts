import { LetterState } from '@/types';

export const NUMBER_OF_ROWS = 6;
export const NUMBER_OF_LETTERS = 5;

export const keyboardRows = [
  ['q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p'],
  ['', 'a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l', ''],
  ['enter', 'z', 'x', 'c', 'v', 'b', 'n', 'm', 'backspace'],
];

export const examples: {
  word: string;
  letterState: LetterState;
  description: string;
}[] = [
  {
    word: 'weary',
    letterState: 'correct',
    description: 'W is in the word and in the correct spot.',
  },
  {
    word: 'pills',
    letterState: 'present',
    description: 'I is in the word but in the wrong spot.',
  },
  {
    word: 'vague',
    letterState: 'absent',
    description: 'G is not in the word in any spot.',
  },
];
