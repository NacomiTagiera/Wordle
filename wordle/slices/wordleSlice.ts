import {
  bindActionCreators,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";

import { useAppDispatch, useAppSelector } from "@/app/hooks";
import { GameState } from "@/types";
import {
  appraiseGuess,
  getRandomWord,
  isGuessComplete,
  isValidWord,
  NUMBER_OF_LETTERS,
  NUMBER_OF_ROWS,
} from "@/lib/word-utils";

const initialState: GameState = {
  boardRowState: [],
  boardState: new Array(NUMBER_OF_ROWS).fill(""),
  currentRowIndex: 0,
  keyboardLetterState: {},
  message: { message: "", duration: 1500 },
  solution: getRandomWord(),
  status: "playing",
};

export const wordleSlice = createSlice({
  name: "wordle",
  initialState,
  reducers: {
    addLetter: (state, action: PayloadAction<string>) => {
      if (state.status !== "playing") return;
      if (state.boardState[state.currentRowIndex].length === NUMBER_OF_LETTERS)
        return;

      state.boardState[state.currentRowIndex] += action.payload;
    },
    removeLetter: (state) => {
      if (state.status !== "playing") return;
      if (!state.boardState[state.currentRowIndex].length) return;

      state.boardState[state.currentRowIndex] = state.boardState[
        state.currentRowIndex
      ].slice(0, -1);
    },
    resetGame: (state) => {
      state.boardRowState = [];
      state.boardState = new Array(NUMBER_OF_ROWS).fill("");
      state.currentRowIndex = 0;
      state.keyboardLetterState = {};
      state.message = { message: "", duration: 1500 };
      state.solution = getRandomWord();
      state.status = "playing";
    },
    resetMessage: (state) => {
      state.message = { message: "", duration: 1500 };
    },
    submitGuess: (state) => {
      const guess = state.boardState[state.currentRowIndex];

      if (!isGuessComplete(guess)) {
        state.message.message = "Not enough letters";
        return;
      }

      if (!isValidWord(guess)) {
        state.message.message = "Not in word list";
        return;
      }

      const boardRowState = appraiseGuess(guess, state.solution);

      state.boardRowState.push(boardRowState);

      boardRowState.forEach((letterState, index) => {
        const currentLetterState = state.keyboardLetterState[guess[index]];

        switch (currentLetterState) {
          case "present":
            if (letterState === "absent") break;

          case "correct":
            break;

          default:
            state.keyboardLetterState[guess[index]] = letterState;
            break;
        }
      });

      if (boardRowState.every((state) => state === "correct")) {
        state.status === "win";
        state.message = { message: "You win!", duration: 0 };
      }

      state.currentRowIndex++;

      if (state.currentRowIndex === state.boardState.length) {
        state.status = "fail";
        state.message = { message: state.solution, duration: 0 };
      }
    },
  },
});

export const useDispatchWordle = () => {
  const { actions } = wordleSlice;
  const dispatch = useAppDispatch();
  return bindActionCreators(actions, dispatch);
};

export const useGetBoardRows = () =>
  useAppSelector((state) =>
    state.boardState.map((letters, rowIndex) => ({
      letters,
      lettersState: state.boardRowState[rowIndex],
    }))
  );

export const useGetGameStatus = () => useAppSelector((state) => state.status);

export const useGetLetterState = () =>
  useAppSelector((state) => state.keyboardLetterState);

export const useGetMessage = () => useAppSelector((state) => state.message);

export default wordleSlice.reducer;
