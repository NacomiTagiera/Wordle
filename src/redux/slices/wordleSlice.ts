import {
  bindActionCreators,
  createSelector,
  createSlice,
  type PayloadAction,
} from '@reduxjs/toolkit';

import { type GameState, type GameStatus, type LetterState } from '@/types';
import { NUMBER_OF_LETTERS, NUMBER_OF_ROWS } from '@/utils/constants';
import { appraiseGuess, getRandomWord, isGuessComplete, isValidWord } from '@/utils/word-utils';

import { useAppDispatch } from '../hooks';

const initialState: GameState = {
  boardRowState: [],
  boardState: new Array(NUMBER_OF_ROWS).fill('') as string[],
  currentRowIndex: 0,
  keyboardLetterState: {} as Record<string, LetterState>,
  message: { message: '', duration: 1500 },
  solution: getRandomWord(),
  status: 'playing' as GameStatus,
};

export const wordleSlice = createSlice({
  name: 'wordle',
  initialState,
  reducers: {
    addLetter: (state, action: PayloadAction<string>) => {
      if (state.status !== 'playing') return;
      if (state.boardState[state.currentRowIndex].length === NUMBER_OF_LETTERS) return;

      state.boardState[state.currentRowIndex] += action.payload;
    },
    removeLetter: (state) => {
      if (state.status !== 'playing') return;
      if (!state.boardState[state.currentRowIndex].length) return;

      state.boardState[state.currentRowIndex] = state.boardState[state.currentRowIndex].slice(
        0,
        -1
      );
    },
    resetGame: (state) => {
      state.boardRowState = [];
      state.boardState = new Array(NUMBER_OF_ROWS).fill('') as string[];
      state.currentRowIndex = 0;
      state.keyboardLetterState = {};
      state.message = { message: '', duration: 1500 };
      state.solution = getRandomWord();
      state.status = 'playing';
    },
    resetMessage: (state) => {
      state.message = { message: '', duration: 1500 };
    },
    submitGuess: (state) => {
      if (state.status !== 'playing') return;

      const guess = state.boardState[state.currentRowIndex];

      if (!isGuessComplete(guess)) {
        state.message.message = 'Not enough letters';
        return;
      }

      if (!isValidWord(guess)) {
        state.message.message = 'Not in word list';
        return;
      }

      const boardRowState = appraiseGuess(guess, state.solution);

      state.boardRowState.push(boardRowState);

      boardRowState.forEach((letterState, index) => {
        const currentLetterState = state.keyboardLetterState[guess[index]];

        if (currentLetterState !== 'present' && currentLetterState !== 'correct') {
          state.keyboardLetterState[guess[index]] = letterState;
        }
      });

      if (state.currentRowIndex === state.boardState.length - 1) {
        state.status = 'fail';
        state.message = { message: state.solution, duration: 0 };
      }

      if (boardRowState.every((state) => state === 'correct')) {
        state.status = 'win';
        state.message = { message: 'You win!', duration: 0 };
      }

      state.currentRowIndex++;
    },
  },
});

export const { addLetter, removeLetter, resetGame, resetMessage, submitGuess } =
  wordleSlice.actions; //for tests

export const useDispatchWordle = () => {
  const { actions } = wordleSlice;
  const dispatch = useAppDispatch();
  return bindActionCreators(actions, dispatch);
};

const selectGameState = (state: GameState) => state;

export const selectBoardRows = createSelector(selectGameState, (state) =>
  state.boardState.map((letters, rowIndex) => ({
    letters,
    lettersState: state.boardRowState[rowIndex],
  }))
);

export const selectGameStatus = (state: GameState) => state.status;

export const selectLetterState = (state: GameState) => state.keyboardLetterState;

export const selectMessage = (state: GameState) => state.message;

export const { reducer } = wordleSlice;
