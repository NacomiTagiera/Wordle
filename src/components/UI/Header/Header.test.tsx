import { fireEvent, screen, waitFor } from '@testing-library/react';

import { initialState } from '@/redux/slices/wordleSlice';
import { type GameState } from '@/types';
import { renderWithProviders } from '@/utils/test-utils';

import { Header } from '.';

describe('Header', () => {
  it('matches the snapshot', () => {
    const { asFragment } = renderWithProviders(<Header />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with text and icons', () => {
    renderWithProviders(<Header />);

    screen.getByLabelText('Help');
    screen.getByLabelText('Restart');
    expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent('Wordle');
  });

  it('opens the instructions when the help button is clicked', async () => {
    renderWithProviders(<Header />);

    fireEvent.click(screen.getByLabelText('Help'));
    await waitFor(() => expect(screen.getByTestId('instructions-container')).toBeVisible());
  });

  it('resets the game when the restart button is clicked', () => {
    const gameState: GameState = {
      ...initialState,
      message: {
        message: 'You win!',
        duration: 500,
      },
      status: 'win',
      currentRowIndex: 5,
    };

    const { store } = renderWithProviders(<Header />, {
      preloadedState: {
        wordle: {
          ...gameState,
        },
      },
    });

    expect(store.getState().wordle).toEqual(gameState);
    fireEvent.click(screen.getByLabelText('Restart'));
    expect(store.getState().wordle).toEqual({
      ...initialState,
      solution: store.getState().wordle.solution,
    });
  });
});
