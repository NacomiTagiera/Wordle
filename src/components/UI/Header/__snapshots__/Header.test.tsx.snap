// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches the snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <button
      aria-label="Help"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium header__icon css-i4bv87-MuiSvgIcon-root"
        data-testid="HelpIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <h1
      class="header__title"
    >
      Wordle
    </h1>
    <button
      aria-label="Restart"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium header__icon css-i4bv87-MuiSvgIcon-root"
        data-testid="RestartAltIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 5V2L8 6l4 4V7c3.31 0 6 2.69 6 6 0 2.97-2.17 5.43-5 5.91v2.02c3.95-.49 7-3.85 7-7.93 0-4.42-3.58-8-8-8zm-6 8c0-1.65.67-3.15 1.76-4.24L6.34 7.34C4.9 8.79 4 10.79 4 13c0 4.08 3.05 7.44 7 7.93v-2.02c-2.83-.48-5-2.94-5-5.91z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </header>
</DocumentFragment>
`;
